### PR TITLE
Adds methods to host DB for batches

### PR DESCRIPTION
**Pull request by joeldudley**
### Why is this change needed?

We need methods for reading and writing batches. These are very similar to the rollup r/w methods.

I also split the `hostdb.go` file between block, rollup and batch files, as it was getting too large. I'd like to do a follow-up PR, creating three separate interfaces for these.

The block and rollup methods are pretty much unchanged, I'd say only the batch stuff needs to be reviewed.

### What changes were made as part of this PR:

- Provide a high level list of the changes made
- List key areas for the reviewer 

### :rotating_light: Definition of done :rotating_light:
- [ ] Unit tests added to cover new or changed functionality 
- [ ] Docs pages updated to cover new or changed functionality
- [ ] [Changelog.md](https://github.com/obscuronet/go-obscuro/blob/main/docs/testnet/changelog.md) updated

### CODE REVIEW

File: go/host/db/batches.go [hunk 1]
```diff
@@ -0,0 +1,112 @@
...
+// DB methods relating to batches.
+
+// headerKey = batchHeaderPrefix  + hash
+func batchHeaderKey(hash gethcommon.Hash) []byte {
+	return append(batchHeaderPrefix, hash.Bytes()...)
+}
+
+// GetHeadBatchHeader returns the header of the node's current head batch, or (nil, false) if no such header is found.
+func (db *DB) GetHeadBatchHeader() (*common.Header, bool) {
+	headBatchHash, found := db.readHeadBatchHash(db.kvStore)
```
Comment (otherview) [inline go/host/db/batches.go:23]: No need to have a method of type and pass a property of type.
```suggestion
	headBatchHash, found := readHeadBatchHash(db.kvStore)
```
OR
```suggestion
	headBatchHash, found := db.readHeadBatchHash()
```

File: go/host/db/batches.go [hunk 2]
```diff
@@ -0,0 +1,112 @@
...
+
+// DB methods relating to batches.
+
+// headerKey = batchHeaderPrefix  + hash
+func batchHeaderKey(hash gethcommon.Hash) []byte {
+	return append(batchHeaderPrefix, hash.Bytes()...)
+}
+
+// GetHeadBatchHeader returns the header of the node's current head batch, or (nil, false) if no such header is found.
+func (db *DB) GetHeadBatchHeader() (*common.Header, bool) {
```
Comment (otherview) [inline go/host/db/batches.go:22]: Optional, but it's likely better to return errors instead of booleans.

File: go/host/db/batches.go [hunk 3]
```diff
@@ -0,0 +1,112 @@
...
+
+// AddBatchHeader adds a batch's header to the known headers
+func (db *DB) AddBatchHeader(header *common.Header, txHashes []common.TxHash) {
+	b := db.kvStore.NewBatch()
+	db.writeBatchHeader(b, header)
+
+	// TODO - #718 - Store the batch txs, batch hash, and batch number per transaction hash, if needed (see `AddRollupHeader`).
+
+	// There's a potential race here, but absolute accuracy of the number of transactions is not required.
+	currentTotal := db.readTotalTransactions(db.kvStore)
```
Comment (otherview) [inline go/host/db/batches.go:38]: Same coment as above

File: go/host/db/batches.go [hunk 4]
```diff
@@ -0,0 +1,112 @@
...
+	db.writeTotalTransactions(b, newTotal)
+
+	// update the head if the new height is greater than the existing one
+	headBatchHeader, found := db.GetHeadBatchHeader()
+	if !found || headBatchHeader.Number.Int64() <= header.Number.Int64() {
+		db.writeHeadBatchHash(b, header.Hash())
+	}
+
+	if err := b.Write(); err != nil {
+		db.logger.Crit("Could not write batch.", log.ErrKey, err)
```
Comment (otherview) [inline go/host/db/batches.go:49]: Should the error be returned here ?

File: go/host/db/batches.go [hunk 5]
```diff
@@ -0,0 +1,112 @@
...
+	}
+}
+
+// Retrieves the batch header corresponding to the hash, or (nil, false) if no such header is found.
+func (db *DB) readBatchHeader(r ethdb.KeyValueReader, hash gethcommon.Hash) (*common.Header, bool) {
+	f, err := r.Has(batchHeaderKey(hash))
+	if err != nil {
+		db.logger.Crit("could not retrieve batch header.", log.ErrKey, err)
```
Comment (otherview) [inline go/host/db/batches.go:57]: Error should be returned

File: go/host/db/batches.go [hunk 6]
```diff
@@ -0,0 +1,112 @@
...
+		db.writeHeadBatchHash(b, header.Hash())
+	}
+
+	if err := b.Write(); err != nil {
+		db.logger.Crit("Could not write batch.", log.ErrKey, err)
+	}
+}
+
+// Retrieves the batch header corresponding to the hash, or (nil, false) if no such header is found.
+func (db *DB) readBatchHeader(r ethdb.KeyValueReader, hash gethcommon.Hash) (*common.Header, bool) {
```
Comment (otherview) [inline go/host/db/batches.go:54]: Optionally, `errutil.ErrNotFound` can be returned for when the resource is not found

File: go/host/db/batches.go [hunk 7]
```diff
@@ -0,0 +1,112 @@
...
+	}
+	header := new(common.Header)
+	if err := rlp.Decode(bytes.NewReader(data), header); err != nil {
+		db.logger.Crit("could not decode batch header.", log.ErrKey, err)
+	}
+	return header, true
+}
+
+// Retrieves the hash of the head batch.
+func (db *DB) readHeadBatchHash(r ethdb.KeyValueReader) (*gethcommon.Hash, bool) {
```
Comment (otherview) [inline go/host/db/batches.go:77]: Error swallowing

File: go/host/db/batches.go [hunk 8]
```diff
@@ -0,0 +1,112 @@
...
+	"github.com/ethereum/go-ethereum/ethdb"
+	"github.com/ethereum/go-ethereum/rlp"
+	"github.com/obscuronet/go-obscuro/go/common"
+	"github.com/obscuronet/go-obscuro/go/common/log"
+)
+
+// DB methods relating to batches.
+
+// headerKey = batchHeaderPrefix  + hash
+func batchHeaderKey(hash gethcommon.Hash) []byte {
```
Comment (otherview) [inline go/host/db/batches.go:17]: Style pref: typically private methods go after the public methods.

File: go/host/db/batches.go [hunk 9]
```diff
@@ -0,0 +1,112 @@
+package db
```
Comment (otherview) [inline go/host/db/batches.go:1]: This should have a unit test file to accompany it. It's nicely written and decoupled should be easy to inject mock objects and test the batches logic

File: go/host/db/blocks.go [hunk 10]
```diff
@@ -0,0 +1,107 @@
...
+	"github.com/ethereum/go-ethereum/core/types"
+	"github.com/ethereum/go-ethereum/ethdb"
+	"github.com/ethereum/go-ethereum/rlp"
+	"github.com/obscuronet/go-obscuro/go/common/log"
+)
+
+// DB methods relating to batches.
+
+// headerKey = blockHeaderPrefix  + hash
+func blockHeaderKey(hash gethcommon.Hash) []byte {
```
Comment (otherview) [inline go/host/db/blocks.go:16]: typically private methods are after the public ones

File: go/host/db/blocks.go [hunk 11]
```diff
@@ -0,0 +1,107 @@
...
+	db.writeBlockHeader(b, header)
+
+	// update the head if the new height is greater than the existing one
+	headBlockHeader, found := db.GetHeadBlockHeader()
+	if !found || headBlockHeader.Number.Int64() <= header.Number.Int64() {
+		db.writeHeadBlockHash(b, header.Hash())
+	}
+
+	if err := b.Write(); err != nil {
+		db.logger.Crit("Could not write rollup.", log.ErrKey, err)
```
Comment (otherview) [inline go/host/db/blocks.go:46]: I don't think we should swallow errors.

File: go/host/db/blocks.go [hunk 12]
```diff
@@ -0,0 +1,107 @@
...
+		db.logger.Crit("could not retrieve head block.", log.ErrKey, err)
+	}
+	if !f {
+		return nil
+	}
+	value, err := r.Get(headBlock)
+	if err != nil {
+		db.logger.Crit("could not retrieve head block.", log.ErrKey, err)
+	}
+	h := gethcommon.BytesToHash(value)
```
Comment (otherview) [inline go/host/db/blocks.go:105]: Can `r.Get(headBlock)` return nil ?

File: go/host/db/rollups.go [hunk 13]
```diff
@@ -0,0 +1,278 @@
...
+
+// GetRollupHeader returns the rollup header given the Hash, or (nil, false) if no such header is found.
+func (db *DB) GetRollupHeader(hash gethcommon.Hash) (*common.Header, bool) {
+	return db.readRollupHeader(db.kvStore, hash)
+}
+
+// AddRollupHeader adds a rollup's header to the known headers
+func (db *DB) AddRollupHeader(header *common.Header, txHashes []common.TxHash) {
+	b := db.kvStore.NewBatch()
+	db.writeRollupHeader(b, header)
```
Comment (otherview) [inline go/host/db/rollups.go:55]: If one write fails should the other rollback ?

File: go/host/db/blocks.go (hunk 11 above)
Comment (joeldudley) [reply go/host/db/blocks.go:46]: Not swallowed - Crit() causes the program to exit.

File: go/host/db/batches.go (hunk 5 above)
Comment (joeldudley) [reply go/host/db/batches.go:57]: Not swallowed - Crit() causes the program to exit.

File: go/host/db/batches.go (hunk 7 above)
Comment (joeldudley) [reply go/host/db/batches.go:77]: Not swallowed - Crit() causes the program to exit.

File: go/host/db/batches.go (hunk 8 above)
Comment (joeldudley) [reply go/host/db/batches.go:17]: Fixed

File: go/host/db/batches.go (hunk 1 above)
Comment (joeldudley) [reply go/host/db/batches.go:23]: Thanks, fixing.

File: go/host/db/blocks.go (hunk 10 above)
Comment (joeldudley) [reply go/host/db/blocks.go:16]: Fixing.

File: go/host/db/batches.go (hunk 5 above)
Comment (otherview) [reply go/host/db/batches.go:56]: Sorry, I mean that the error isn't being returned, swallowed was the wrong word, but it was in the sense that the caller doesn't have to handle it.

Comment (joeldudley) [reply go/host/db/batches.go:56]: Yeah, the approach we’ve chosen here is for the process to exit if the db is unreachable or has a critical issue. Maybe I can add a todo to handle critical db errors gracefully?

File: go/host/db/blocks.go (hunk 12 above)
Comment (joeldudley) [reply go/host/db/blocks.go:105]: Only if the err is not nil, which we check.

File: go/host/db/rollups.go (hunk 13 above)
Comment (joeldudley) [reply go/host/db/rollups.go:35]: I think so. Do you think we shouldn't?

File: go/host/db/batches.go (hunk 2 above)
Comment (joeldudley) [reply go/host/db/batches.go:16]: Raised https://github.com/obscuronet/obscuro-internal/issues/1195. Will tackle after.

File: go/host/db/batches.go (hunk 4 above)
Comment (joeldudley) [reply go/host/db/batches.go:43]: Raised https://github.com/obscuronet/obscuro-internal/issues/1195. Will tackle after.

File: go/host/db/batches.go (hunk 5 above)
Comment (joeldudley) [reply go/host/db/batches.go:56]: Raised https://github.com/obscuronet/obscuro-internal/issues/1195. Will tackle after.

File: go/host/db/batches.go (hunk 7 above)
Comment (joeldudley) [reply go/host/db/batches.go]: Raised https://github.com/obscuronet/obscuro-internal/issues/1195. Will tackle after.

File: go/host/db/blocks.go (hunk 11 above)
Comment (joeldudley) [reply go/host/db/blocks.go:41]: Raised https://github.com/obscuronet/obscuro-internal/issues/1195. Will tackle after.

File: go/host/db/rollups.go (hunk 13 above)
Comment (otherview) [reply go/host/db/rollups.go:35]: I think we should. But it seems like at the moment if one write fails the approach is to panic. Is that intended yeah ?

**Review by otherview — APPROVED** [review]
I think there are a few todos on unit testing and error handling, other than that lgtm 👍

File: go/host/db/rollups.go (hunk 13 above)
Comment (joeldudley) [reply go/host/db/rollups.go:35]: Yeah - you want all the writes to succeed or fail together.

Comment (otherview) [reply go/host/db/rollups.go:35]: I understand that's implicit, but what if there was a TODO mentioning that ? At the moment if one write fails the whole node stops, it's not an explicit logic of what its expected to happen.

Comment (joeldudley) [reply go/host/db/rollups.go:35]: That'll be picked up by the follow-up ticket, if that's ok

File: go/host/db/blocks_test.go [hunk 14]
```diff
@@ -0,0 +1,89 @@
+package db
+
+import (
+	"math/big"
+	"testing"
+
+	"github.com/ethereum/go-ethereum/core/types"
+)
+
+const magicNumber = 777
```
Comment (otherview) [inline go/host/db/blocks_test.go:10]: Gonna guess the magic number is the minimum required for making a header ? Can a comment be added to it ?

**Review by otherview — APPROVED** [review]
Would add a couple of comments, lgtm

File: go/host/db/blocks_test.go (hunk 14 above)
Comment (joeldudley) [reply go/host/db/blocks_test.go:10]: No, it's just a field to check is still there when the object comes back from the DB. Will rename.

**Pull request merged**